### DISCUSSION TITLE
Default filter predicates

### DISCUSSION THREAD

**Discussion by marcflp** [trailblazer/trailblazer-finder#28] [Ideas]
As discussed on slack, konung would like to see the following:
> I'd like it to allow me to encode searching into request param: 
-  I can do things like example.com/posts?author_id_eq=1&title_startswith=“Konung”&editor_last_name_or_first_name_cont=“Gorb”

**Comment by konung** [comment]
This idea is blatantly stolen from ransack :) 
 https://github.com/activerecord-hackery/ransack#search-matchers .

 It doesn't have to be the same - but it's a good starting point

**Comment by marcflp** [comment]
We'll start with a few basic ones and go from there:
- equal / not equal
- matches / does not match
- less than / greater than
- less than or equal / greater than or equal
- present / not present
- true / false

These would only be available on fields that have been specified with `filter_by :field_name` without any custom methods/blocks for starters.

**Comment by apotonick** [comment]
Is ransack AR-only? Could we use parts of their logic here?

**Comment by marcflp** [comment]
Ransack is AR only yeah, we'll have to find a way to do it different

**Comment by konung** [comment]
Yes it is. But here is something I found for: Sequel - https://github.com/djellemah/philtre

**Comment by marcflp** [comment]
I think i'll reuse part of their predicate splitter logic @konung

**Comment by apotonick** [comment]
It would be cool if we can add them as a dependency and help opening their API so we can reuse as much as possible.

**Comment by marcflp** [comment]
We really can't, `ransack` is AR only and `philtre` is Sequel only.

Unless you want to add both as dependency for either case, and translate towards their respective predicates (bit much if you ask me, cause that would equal to doing it ourselves)

**Comment by apotonick** [comment]
No of course not :joy:

**Comment by marcflp** [comment]
I figured you'd say that ;)

I'll come up with a PoC for our own solution for this in the next few days

**Comment by marcflp** [comment]
@konung https://github.com/trailblazer/trailblazer-finder/releases/tag/0.1.2

Initial release for default predicate filters

**Comment by konung** [comment]
Copy from slack for consistency sake:

yep just got the notification.

I want to say : https://github.com/activerecord-hackery/ransack/blob/master/lib/ransack/locale/en.yml#L15

But to be honest we only use about 1/2 of those
the other often used ones by us are `starts` (‘aka startswith’), cont ( “contains”) and matches.   They could all probably be replaced with something called “matches” that maps to SQL “LIKE ” and “NOT LIKE”.
Going even further than that they could all probably be replaced with “REGEXP” or “RLIKE”  ( although I think it’s something slightly different in PostgreSQL - I use pretty much only MariaDB)
Based on a brief search on google it seems like every major DB vendor has slightly different support for Rexexp.
“LIKE” & “NOT LIKE” seem to be universal enough, that maybe it’s good enough for now.
Certainly would work for my use case

**Comment by konung** [comment]
One more we use sometimes is: `_in`  which translates to SQL: SELECT * FROM mytable IN ('value1', 'value2', 'value3')

**Comment by marcflp** [comment]
# Update
@fernandes kindly offered to implement the other predicates in a feature branch, while i'm going to focus on multiple order columns and rom-rb adapter.

## So just for reference
@fernandes here's a little text on what we still need (and anything else that might come up in discussions on slack), and where to find what for predicates
### We need the following predicates still (for Hash, AR and Sequel), and their opposites:
- sw / not_sw - starts with / does not start with : LIKE 'value%'
- ew / not_ew - ends with / does not end with : LIKE '%value'
- cont / not_cont - contains / does not contain: LIKE '%value%'
- in / not_in: IN ('value1', 'value2', 'value3')

### Files for Predicates
- Predicates for Hash are located at: 
[lib/trailblazer/finder/predicates.rb](https://github.com/trailblazer/trailblazer-finder/blob/master/lib/trailblazer/finder/predicates.rb)
- Predicates for ActiveRecord are located at: 
[lib/trailblazer/finder/adapters/active_record/predicates.rb](https://github.com/trailblazer/trailblazer-finder/blob/master/lib/trailblazer/finder/adapters/active_record/predicates.rb)
- Predicates for Sequel are located at: 
[lib/trailblazer/finder/adapters/sequel/predicates.rb](https://github.com/trailblazer/trailblazer-finder/blob/master/lib/trailblazer/finder/adapters/sequel/predicates.rb)
- Predicates for DataMapper (if you get some time) are to placed in: 
[lib/trailblazer/finder/adapters/data_mapper/](https://github.com/trailblazer/trailblazer-finder/tree/master/lib/trailblazer/finder/adapters/data_mapper)

### Tests for Predicates:
- Finder Macro test with Predicates / ActiveRecord: 
[spec/trailblazer/operation/predicates_spec.rb](https://github.com/trailblazer/trailblazer-finder/blob/master/spec/trailblazer/operation/predicates_spec.rb)
- Features test with Predicates: 
[spec/trailblazer/finder/features/predicates_spec.rb](https://github.com/trailblazer/trailblazer-finder/blob/master/spec/trailblazer/finder/features/predicates_spec.rb)
- ActiveRecord adapter test with Predicates: 
[spec/trailblazer/finder/adapters/active_record/predicates_spec.rb](https://github.com/trailblazer/trailblazer-finder/blob/master/spec/trailblazer/finder/adapters/active_record/predicates_spec.rb)
- Sequel adapter test with Predicates: 
[spec/trailblazer/finder/adapters/sequel/predicates_spec.rb](https://github.com/trailblazer/trailblazer-finder/blob/master/spec/trailblazer/finder/adapters/sequel/predicates_spec.rb)

**Comment by konung** [comment]
1. What about negative matchers for `NOT LIKE `:
* _not_starts_with
* _not_ends_with
* _not_contains
* _not_in ( `NOT IN`)
2. Also what about RLIKE / REGEXP / REGEX_MATCHES? Or would it be too hard to support across various Databases?

**Comment by marcflp** [comment]
@konung was already updating previous message. Regex shouldn't be too difficult either, but let's get these done first, then we move on to those.

**Comment by marcflp** [comment]
With https://github.com/trailblazer/trailblazer-finder/issues/8 and https://github.com/trailblazer/trailblazer-finder/commit/c2fa669c47d90cd1086ac5e93983809a471983ba DataMapper now also has default predicates.

All that's left for 0.2.0 release would be the predicates mentioned up here and ROM-RB Adapter.

**Comment by marcflp** [comment]
@fernandes i added sw, ew, cont and their opposites for AR and some basic tests to cover them

**Comment by marcflp** [comment]
New predicates we should make:
at / before / after for dates/times
